---
 arch/powerpc/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.24.3-rt3/arch/powerpc/Kconfig
===================================================================
--- linux-2.6.24.3-rt3.orig/arch/powerpc/Kconfig	2008-02-26 23:29:37.000000000 -0500
+++ linux-2.6.24.3-rt3/arch/powerpc/Kconfig	2008-02-26 23:30:03.000000000 -0500
@@ -79,6 +79,7 @@ config ARCH_NO_VIRT_TO_BUS
 config PPC
 	bool
 	default y
+	select HAVE_MCOUNT
 
 config EARLY_PRINTK
 	bool
