---
 arch/x86/kernel/io_apic_64.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.24.7-rt21/arch/x86/kernel/io_apic_64.c
===================================================================
--- linux-2.6.24.7-rt21.orig/arch/x86/kernel/io_apic_64.c	2008-10-08 22:22:41.000000000 -0400
+++ linux-2.6.24.7-rt21/arch/x86/kernel/io_apic_64.c	2008-10-08 22:23:21.000000000 -0400
@@ -1440,7 +1440,8 @@ static void ack_apic_level(unsigned int 
 	irq_complete_move(irq);
 #if defined(CONFIG_GENERIC_PENDING_IRQ) || defined(CONFIG_IRQBALANCE)
 	/* If we are moving the irq we need to mask it */
-	if (unlikely(irq_desc[irq].status & IRQ_MOVE_PENDING)) {
+	if (unlikely(irq_desc[irq].status & IRQ_MOVE_PENDING) &&
+	    !(irq_desc[irq].status & IRQ_INPROGRESS)) {
 		do_unmask_irq = 1;
 		mask_IO_APIC_irq(irq);
 	}
