---
 kernel/sched_cpupri.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24.7-rt21/kernel/sched_cpupri.h
===================================================================
--- linux-2.6.24.7-rt21.orig/kernel/sched_cpupri.h	2008-10-08 22:23:05.000000000 -0400
+++ linux-2.6.24.7-rt21/kernel/sched_cpupri.h	2008-10-08 22:23:59.000000000 -0400
@@ -12,7 +12,7 @@
 /* values 2-101 are RT priorities 0-99 */
 
 struct cpupri_vec {
-	spinlock_t lock;
+	raw_spinlock_t lock;
 	int        count;
 	cpumask_t  mask;
 };
