

Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
---
 kernel/lockdep.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.24.7-rt27/kernel/lockdep.c
===================================================================
--- linux-2.6.24.7-rt27.orig/kernel/lockdep.c	2009-02-08 00:02:01.000000000 -0500
+++ linux-2.6.24.7-rt27/kernel/lockdep.c	2009-02-08 00:03:01.000000000 -0500
@@ -2757,6 +2757,9 @@ lock_set_subclass(struct lockdep_map *lo
 {
 	unsigned long flags;
 
+	if (unlikely(!lock_stat && !prove_locking))
+		return;
+
 	if (unlikely(current->lockdep_recursion))
 		return;
 
