---
 arch/x86/kernel/nmi_32.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24.3-rt3/arch/x86/kernel/nmi_32.c
===================================================================
--- linux-2.6.24.3-rt3.orig/arch/x86/kernel/nmi_32.c	2008-02-26 23:30:52.000000000 -0500
+++ linux-2.6.24.3-rt3/arch/x86/kernel/nmi_32.c	2008-02-26 23:31:02.000000000 -0500
@@ -392,7 +392,7 @@ nmi_watchdog_tick(struct pt_regs *regs, 
 
 		spin_lock(&lock);
 		printk("NMI backtrace for cpu %d\n", cpu);
-		dump_stack();
+		show_regs(regs);
 		spin_unlock(&lock);
 		cpu_clear(cpu, backtrace_mask);
 	}
