---
 init/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.24.7-rt21/init/Kconfig
===================================================================
--- linux-2.6.24.7-rt21.orig/init/Kconfig	2008-10-08 22:24:15.000000000 -0400
+++ linux-2.6.24.7-rt21/init/Kconfig	2008-10-08 22:24:23.000000000 -0400
@@ -647,6 +647,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
