Subject: sched-fix-dequeued-race.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 24 Jul 2008 16:14:31 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/sched.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24.7-rt27/kernel/sched.c
===================================================================
--- linux-2.6.24.7-rt27.orig/kernel/sched.c	2009-02-08 00:05:00.000000000 -0500
+++ linux-2.6.24.7-rt27/kernel/sched.c	2009-02-08 00:05:05.000000000 -0500
@@ -3769,7 +3769,7 @@ void scheduler_tick(void)
 		rq->clock = next_tick;
 	rq->tick_timestamp = rq->clock;
 	update_cpu_load(rq);
-	if (curr != rq->idle) /* FIXME: needed? */
+	if (curr != rq->idle && curr->se.on_rq)
 		curr->sched_class->task_tick(rq, curr);
 	spin_unlock(&rq->lock);
 
