From: Steven Rostedt <srostedt@redhat.com>
Subject: ftrace: fix the preempt-disable start fix

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 kernel/trace/trace_irqsoff.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.24.7-rt21/kernel/trace/trace_irqsoff.c
===================================================================
--- linux-2.6.24.7-rt21.orig/kernel/trace/trace_irqsoff.c	2008-10-08 22:25:09.000000000 -0400
+++ linux-2.6.24.7-rt21/kernel/trace/trace_irqsoff.c	2008-10-08 22:25:19.000000000 -0400
@@ -364,9 +364,9 @@ void trace_preempt_on(unsigned long a0, 
 
 void trace_preempt_off(unsigned long a0, unsigned long a1)
 {
-	start_critical_timing(a0, a1);
+	tracing_hist_preempt_start();
 	if (preempt_trace())
-		tracing_hist_preempt_start();
+		start_critical_timing(a0, a1);
 }
 #endif /* CONFIG_PREEMPT_TRACER */
 
