#! /bin/sh /usr/share/dpatch/dpatch-run
diff --git a/arch/x86/kernel/reboot_32.c b/arch/x86/kernel/reboot_32.c
index bb1a0f8..50c2f11 100644
--- a/arch/x86/kernel/reboot_32.c
+++ b/arch/x86/kernel/reboot_32.c
@@ -135,6 +135,16 @@ static struct dmi_system_id __initdata reboot_dmi_table[] = {
 			DMI_MATCH(DMI_PRODUCT_NAME, "HP Compaq"),
 		},
 	},
+	{	/* Handle problems with rebooting on Intel Crown Beach board */
+		.callback = set_bios_reboot,
+		.ident = "Intel Crown Beach board",
+		.matches = {
+			/* Currently the DMI info is not customized and indicates 
+                         * OEM need change that */
+			DMI_MATCH(DMI_SYS_VENDOR, "To Be Filled By O.E.M."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "To Be Filled By O.E.M."),
+		},
+	},
 	{ }
 };
 
