From 54c5b7426aae1d374df4c7ab6adaeba4e40a51d3 Mon Sep 17 00:00:00 2001
From: Alexey Dobriyan <adobriyan@openvz.org>
Date: Tue, 18 Mar 2008 12:07:46 +0300
Subject: [PATCH 41/48] Linux 2.6.24-ovz003

---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

Index: kernel/Makefile
===================================================================
--- kernel.orig/Makefile	2008-11-24 15:47:45.000000000 +0100
+++ kernel/Makefile	2008-11-24 15:59:06.000000000 +0100
@@ -3,7 +3,7 @@
 SUBLEVEL = 24
 EXTRAVERSION = .6
 NAME = Err Metey! A Heury Beelge-a Ret!
-VZVERSION = ovz002
+VZVERSION = ovz003
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
