From 483a70883c85acf684ffbd8fa6a4c9bfade878f3 Mon Sep 17 00:00:00 2001
From: Alexey Dobriyan <adobriyan@openvz.org>
Date: Thu, 8 May 2008 13:04:01 +0400
Subject: [PATCH 73/73] Linux 2.6.24-ovz005

---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

Index: kernel/Makefile
===================================================================
--- kernel.orig/Makefile	2008-11-24 15:59:19.000000000 +0100
+++ kernel/Makefile	2008-11-24 16:00:18.000000000 +0100
@@ -3,7 +3,7 @@
 SUBLEVEL = 24
 EXTRAVERSION = .6
 NAME = Err Metey! A Heury Beelge-a Ret!
-VZVERSION = ovz004
+VZVERSION = ovz005
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
